### PR TITLE
Test perf buttwoo

### PR DESCRIPTION
**Pull request by arj03**
```
create 5000 new messages: 1.242s
validate 5000 messages ssb-buttwoo: 517.988ms
validate 5000 messages sbot: 585.584ms
native to db format 5000 messages ssb-buttwoo: 166.615ms
native to db format 5000 messages sbot: 128.726ms
db to native format 5000 messages ssb-buttwoo: 134.514ms
db to native format 5000 messages sbot: 193.874ms
```

### CODE REVIEW

**Comment by arj03** [comment]
This is written a bit quickly. I don't have so much time today. But wanted to put it here as it can make it easier to figure out why EBT is not as fast as before.

db to native format 5000 messages sbot: 193.874ms <- a lot of this overhead is finding the format (around 40-50ms)

**Comment by arj03** [comment]
The latest commit should better match what is going on when replicating over ebt:

```
create 5000 new messages: 1.204s
validate 5000 messages ssb-buttwoo: 509.465ms
validate 5000 messages sbot: 587.652ms
native to db format 5000 messages ssb-buttwoo: 164.59ms
native to db format 5000 messages sbot: 149.172ms
db to native format 5000 messages ssb-buttwoo: 138.734ms
db to native format 5000 messages sbot: 188.189ms
```

**Comment by staltz** [comment]
@arj03 These benchmarks were useful, I found out that too much work was done with `SSBURI.decompose` and that basically in many cases we were converting from BFE to URI and back to BFE, so I made quicker versions which just do the thing directly.

Now validation with feed-formats/buttwoo-v1 is slightly faster than validation in ssb-buttwoo:

**Run 1**

```
create 5000 new messages: 1.088s
validate 5000 messages ssb-buttwoo: 522.733ms
validate 5000 messages sbot: 473.532ms
```

**Run 2**

```
create 5000 new messages: 944.712ms
validate 5000 messages ssb-buttwoo: 460.944ms
validate 5000 messages sbot: 463.95ms
```

**Run 3**

```
create 5000 new messages: 997.205ms
validate 5000 messages ssb-buttwoo: 470.361ms
validate 5000 messages sbot: 412.971ms
```

I put the commit https://github.com/ssbc/ssb-db2/pull/351/commits/4c3e0c7133de5b638ed8cff1a2862a75bb56b9c6 here in this PR. We can merge at some point, but thought it would make sense to put all the performance stuff (measurements and optimizations) here first.

**Comment by staltz** [comment]
Another optimization: https://github.com/ssbc/ssb-db2/pull/351/commits/f9af0005580e9ba8ef64c15af606a2db5f3567b4 I copy-pasted ssb-buttwoo's bipfToButt2 internals, and that made a bit faster.

It's still slower, but there are two things to be done. bipfToButt2 seems slightly incorrect because the `signature` from the database is a buffer, when it should be a string. So my feed-formats/buttwoo-v1 is doing one conversion of string to buffer that bipfToButt2 is not doing. And then the second thing is that `db.findFeedFormatForAuthor(author)` going on, which I think we can get rid of if we pipe the feed format along the function calls.

```
create 5000 new messages: 946.073ms
validate 5000 messages ssb-buttwoo: 411.721ms
validate 5000 messages sbot: 406.933ms
native to db format 5000 messages ssb-buttwoo: 143.907ms
native to db format 5000 messages sbot: 110.885ms
db to native format 5000 messages ssb-buttwoo: 114.169ms
db to native format 5000 messages sbot: 130.072ms
```

**Comment by staltz** [comment]
Alright, now with commit https://github.com/ssbc/ssb-db2/pull/351/commits/6027e09979c0e15a1f104f4ad3994c324200e830 all measurements are faster in the new implementation.

**Run 1**

```
create 5000 new messages: 933.821ms
validate 5000 messages ssb-buttwoo: 444.749ms
validate 5000 messages sbot: 418.946ms
native to db format 5000 messages ssb-buttwoo: 159.472ms
native to db format 5000 messages sbot: 129.618ms
db to native format 5000 messages ssb-buttwoo: 134.006ms
db to native format 5000 messages sbot: 123.607ms
```

**Run 2**

```
create 5000 new messages: 923.147ms
validate 5000 messages ssb-buttwoo: 442.23ms
validate 5000 messages sbot: 426.279ms
native to db format 5000 messages ssb-buttwoo: 160.077ms
native to db format 5000 messages sbot: 125.565ms
db to native format 5000 messages ssb-buttwoo: 131.653ms
db to native format 5000 messages sbot: 121.745ms
```

**Run 3**

```
create 5000 new messages: 965.35ms
validate 5000 messages ssb-buttwoo: 440.03ms
validate 5000 messages sbot: 413.354ms
native to db format 5000 messages ssb-buttwoo: 161.078ms
native to db format 5000 messages sbot: 127.85ms
db to native format 5000 messages ssb-buttwoo: 128.372ms
db to native format 5000 messages sbot: 118.041ms
```

File: log.js [hunk 1]
```diff
@@ -96,26 +96,39 @@ module.exports = function (dir, config, privateIndex, db) {
...
+      let format
       if (!feedFormat) {
+        const pValueAuthor = bipf.seekKey2(buffer, pValue, BIPF_AUTHOR, 0)
+        const author = bipf.decode(buffer, pValueAuthor)
+        format = db.findFeedFormatForAuthor(author)
+        if (!format) {
+          // prettier-ignore
+          return cb(new Error('getNativeMsg() failed because this author is for an unknown feed format: ' + author))
+        }
+      } else if (typeof format === 'string') {
```
Comment (arj03) [inline log.js:114]: this line needs to be feedFormat === 'string'

**Comment by arj03** [comment]
This is certainly better, in ebt I'm at around 6800. So there is still something there. I'll have a proper look tonight or tomorrow. Think we are close, there has to be something else these benchmarks don't show, but at least there where some good optimizations here.

File: log.js (hunk 1 above)
Comment (staltz) [reply log.js]: Fixed!

**Comment by staltz** [comment]
@arj03 I have another idea. In the log, all the BIPF messages have the same "schema", regardless of feed format. This is because we need to do that because we use BIPF seeking for the indexes. But the `signature` field is never used for BIPF seeking, nor queries nor indexes. The signature field is only relevant for the feed format, during validation. So we could allow the `signature` field of a BIPF-ified buttwoo (i.e. `fromNativeMsg(nativeMsg, 'bipf')`) to be a buffer, and this means we can skip all the "string signature to buffer signature" conversions, and gain in perf.

**Comment by staltz** [comment]
> This is certainly better, in ebt I'm at around 6800.

Also, (I'm sure you know about this trick, but if we are lucky, you forgot about it) make sure to delete node_modules and reinstall because ssb-db2 is installed via a git branch and gotta make sure the right commit was downloaded. Just in case

**Comment by arj03** [comment]
> re signature

Yes. ssb-buttwoo stored is as a buffer because of this. We anyway store things like tag and contentHash as a buffer, so if we want a pretty printer for view source / txt export, it has to handle those fields anyway.

**Comment by arj03** [comment]
That said, I don't think it is what the difference is. Maybe 100 more

**Comment by staltz** [comment]
@arj03 Can we merge this PR into formats-split? (PS: I added some new measurement utils, `startMeasure`/`endMeasure` to the benchmark script)

**Comment by arj03** [comment]
Sure

**Pull request merged**